src/coreclr/jit/importer.cpp

```diff
@@ -5278,6 +5270,89 @@ GenTree* Compiler::impArrayAccessIntrinsic(
+// optimization, if the object to keep alive is a GT_BOX, as a series
+// of statements (prepended as needed) with GT_KEEPALIVE top-level nodes,
+// for each field of a reference type in the struct's layout. As part
+// of this, removes the box's allocation and inserts a copy of the
+// box's source to a temporary, if the source wasn't a local already,
+// to enable the re-use of this local for field indirections.
+//
+// Arguments:
+//    objToKeepAlive - the intrinisic call's argument
+//
+// Return Value:
+//    The imported GT_KEEPALIVE, with "objToKeepAlive" as the operand.
+//    If the GT_BOX optimization was successful, always returns a GT_NOP.
+//
+GenTree* Compiler::impKeepAliveIntrinsic(GenTree* objToKeepAlive)
+{
+    assert(objToKeepAlive->TypeIs(TYP_REF));
+
+    if (opts.OptimizationEnabled() && objToKeepAlive->IsBoxedValue())
+    {
+        CORINFO_CLASS_HANDLE boxedClass = lvaGetDesc(objToKeepAlive->AsBox()->BoxOp()->AsLclVar())->lvClassHnd;
+        GenTree*             boxSrc = gtTryRemoveBoxUpstreamEffects(objToKeepAlive, Compiler::BR_REMOVE_BUT_NOT_NARROW);
+
+        if (boxSrc != nullptr)
+        {
+            JITDUMP("\nImporting KEEPALIVE(BOX) as a series of KEEPALIVE(FIELD)\n");
+            INDEBUG(bool foundGcField = false);
+
+            unsigned boxTempNum;
+            if (boxSrc->OperIsLocal())
+            {
+                boxTempNum = boxSrc->AsLclVarCommon()->GetLclNum();
+            }
+            else
+            {
+                boxTempNum            = lvaGrabTemp(true DEBUGARG("Temp for the box source"));
+                GenTree*   boxTempAsg = gtNewTempAssign(boxTempNum, boxSrc);
+                Statement* boxAsgStmt = objToKeepAlive->AsBox()->gtCopyStmtWhenInlinedBoxValue;
+                boxAsgStmt->SetRootNode(boxTempAsg);
+            }
+
+            ClassLayout* layout = typGetObjLayout(boxedClass);
+            if (layout->HasGCPtr())
+            {
+                for (unsigned slot = 0; slot < layout->GetSlotCount(); slot++)
+                {
+                    if (layout->IsGCPtr(slot))
+                    {
+                        // This is a very verbose way of saying gtNewLclFldNode.
+                        // However, gtNewLclFldNode will not work for implicit byrefs.
+                        // TODO-Addr: replace this with a method that accounts for this.
+
+                        GenTree* fieldOffset = gtNewIconNode(slot * TARGET_POINTER_SIZE);
+                        GenTree* boxTemp     = gtNewLclvNode(boxTempNum, boxSrc->TypeGet());
+                        GenTree* boxSrcAddr  = gtNewOperNode(GT_ADDR, TYP_BYREF, boxTemp);
+                        GenTree* fieldAddr   = gtNewOperNode(GT_ADD, TYP_BYREF, boxSrcAddr, fieldOffset);
+                        GenTree* field       = gtNewIndir(TYP_REF, fieldAddr);
+                        GenTree* keepalive   = gtNewKeepAliveNode(field);
+
+                        Statement* stmt = gtNewStmt(keepalive, impCurStmtOffs);
```

Comment:
I expect such structs to always be on the stack (because they have GC fields and used as a whole, so no independent promotion) so we don't have precise fields' liveness for them, we do zero-init and always report them as alive, don't we? 

So my theory for such test:
```
S s = struct with GC fields;
nativeCallThatTakesStructAndDoesSomethingWithGCField(s);
GC.KeepAlive(s);
```
it will currently work even if you delete `GC.KeepAlive(s)` just because the struct `S` will be allocated on the stack and will be reported as live for the whole method. The only exception, in theory, could be a struct with 1 byref field
```
struct S
{
  Object o;
}
```
in this case Jit can replace the struct with its field and do precise liveness for it.
Am I correct and it will work even without KeepAlive ?


However, the previous example is just a current limitation.

What I don't understand is why we can't do this:
```
if (layout->HasGCPtr())
{
  GenTree* boxTemp     = gtNewLclvNode(boxTempNum, boxSrc->TypeGet());
  GenTree* boxSrcAddr  = gtNewOperNode(GT_ADDR, TYP_BYREF, boxTemp);
  GenTree* keepalive = gtNewKeepAliveNode(boxSrcAddr);
  addStmt.
}
```
and keep one GT_KEEPALIVE for the stuct, not for each GC field?

Reply:
So after thinking about this and some testing, I've come to the conclusion that the concern that fields in structs in foreign frames won't be "kept alive" is "real", but also per-existing and not something that much can (should) be done about.

Structs in the frame of the method being compiled will be kept alive by virtue of being address-exposed if we take the address of the whole struct, while implicit byrefs will be kept alive by their owner method.

In terms of CQ, the current approach of field-by-field importation produces sequences of dummy loads into a register like `mov rax, [fld1]`, `mov rax, [fld2]`, etc. Address-exposing the struct will avoid that, but will obviously impede other possible optimizations, though presumably that will matter little. It does seem a little weird to generate a dummy use of a byref just so that the struct is marked address-exposed, thus indirectly making it live for the whole method, but perhaps that's by design.

Reply:
I feel like I am a little behind, how do we get `address exposed` set for the struct? Do we set `address exposed` when using your current approach with fields?

Reply:
> I feel like I am a little behind, how do we get address exposed set for the struct? Do we set address exposed when using your current approach with fields?

So if we use the address of the whole struct, it is as if we passed it to a call as an argument - that is why we'll have it address-exposed.

> Do we set address exposed when using your current approach with fields?

It appears so...
`LocalAddressVisitor` sets it, and only later does morph fold it back to a `LCL_FLD`. I initially got confused between the two :(. Given this fact, I see no advantage to keeping the field approach and will use the whole struct as you suggested.

Reply:
> LocalAddressVisitor sets it, and only later does morph fold it back to a LCL_FLD. I initially got confused between the two :(. Given this fact, I see no advantage to keeping the field approach and will use the whole struct as you suggested.

It is not important when we fold them, because if the whole struct is marked as address exposed we can't do anything with the fields even if they are all in `LCL_FLD` form.

> So if we use the address of the whole struct, it is as if we passed it to a call as an argument - that is why we'll have it address-exposed.

what if the struct is passed by value?

Reply:
> It is not important when we fold them, because if the whole struct is marked as address exposed we can't do anything with the fields even if they are all in LCL_FLD form.

Yes I understand. I just thought that it was the address visitor that folded them (and didn't set the address exposed flag in doing so), for some reason.

> what if the struct is passed by value?

You mean to `KEEPALIVE`? That would require codegen support, wouldn't it?

Reply:
no, if the struct is 8-byte size or 16-byte on x64 unix/arm and it is passed to unmanaged call by value, is it marked as address exposed?

Reply:
No, as you say, it is not. I think I misunderstood your original question...